draft-ietf-quic-transport.md

```diff
@@ -1224,41 +1225,77 @@ be routed back to the endpoint and identified by the endpoint upon receipt.
-or received on that local address.  This ensures that when an endpoint migrates
-to a new path or changes connection ID on an existing path, the packets will use
-a new connection ID in both directions.
-
-Implementations SHOULD ensure that peers have a connection ID with a matching
-sequence number available when changing to a new connection ID.  An
-implementation could do this by always supplying a corresponding connection ID
-to a peer for each connection ID received from that peer.
-
-While endpoints select connection IDs as appropriate for their implementation,
-the connection ID MUST NOT include the unprotected sequence number.  Endpoints
-need to be able to recover the sequence number associated with each connection
-ID they generate without relying on information available to unaffiliated
-parties. A connection ID that encodes an unencrypted sequence number could be
-used to correlate connection IDs across network paths.
+An endpoint whose peer has selected a non-zero-length connection ID will
+potentially send packets using different connection IDs over the lifetime of a
+connection. These connection IDs are supplied by the peer using the
+NEW_CONNECTION_ID frame ({{frame-new-connection-id}}).
+
+
+### Issuing Connection IDs
+
+An endpoint issues connection IDs to its peer for the peer to use when sending
+packets. This allows the endpoint control over the strategy used to interpret
+connection IDs of packets that it receives.  These connection IDs are
+communicated to the peer using NEW_CONNECTION_ID frames
+({{frame-new-connection-id}}).
+
+The initial connection ID issued to the peer is sent by the endpoint as the
+Source Connection ID during the handshake, ensuring that the peer always begins
+the connection with at least one connection ID to use when sending.  An endpoint
+SHOULD supply its peer with additional connection IDs via NEW_CONNECTION_ID
+frames. While each endpoint can choose how many connection IDs to issue, a
+recommended number of outstanding connection IDs is eight.
+
+When an endpoint issues a connection ID, it MUST accept packets using this
+connection ID for the duration of the connection or until its peer invalidates
+the connection ID via a CONNECTION_ID_FINISHED frame
+({{frame-connection-id-finished}}).
+
+Implementations SHOULD ensure that peers have sufficient connection IDs
+available to reduce the possibility of peers exhausting their supply of
+available connection IDs.  An implementation could do this by always supplying a
+new connection ID for each connection ID retired with a CONNECTION_ID_FINISHED
+frame. When a receiver of a packet notices that its peer is now using a
+previously unused connection ID, it can choose to supply its peer with a new
+connection ID using a NEW_CONNECTION_ID frame to reduce the possibility of its
+peer running out of available connection IDs.
+
+An endpoint that receives a packet with a different remote address than
```

Comment:
I think theoretically we should require the peer to switch to a new CID on an existing 5-tuple when an endpoint switches to a new one.

Such behavior is beneficial to protect the lifetimes of connections between servers. It is natural to expect many QUIC connections between servers to share a single 5-tuple. In such case, an endpoint might want to protect the lifetime of each connection by randomly switching to a new CID. But unless we require CID migration on an existing path to be complementary, you cannot protect it...

Reply:
So we have: 
> An endpoint that receives a packet with a different remote address or with a different connection ID than previously used SHOULD also switch to sending with a different connection ID.

I don't think it can be stronger than a SHOULD because the other endpoint may not have enough CIDs to actually switch.

Reply:
More to the point, if you require a change, you create the potential for infinite regress.  "They changed, so I will change" can repeat until you run out of connection IDs.

Reply:
Oh, and one model that was suggested was greasing of connection ID use.  I could send with another connection ID 1 out of N packets.  That could trigger exhaustion as well.